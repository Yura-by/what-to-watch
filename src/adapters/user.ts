class User {
  id: number;
  email: string;
  name: string;
  avatarUrl: string;

  constructor(raw) {
    this.id = raw.id;
    this.email = raw.email;
    this.name = raw.name;
    this.avatarUrl = raw.avatar_url;
  }
}

export default User;
