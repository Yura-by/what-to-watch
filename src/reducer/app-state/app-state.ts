import {Action} from './../../types';
import {Genre} from '../../const';

interface State {
  genre: string;
  selectedMovie: number;
}

const initialState: State = {
  genre: Genre.ALL_GENRES,
  selectedMovie: 1,
};

enum ActionType {
  SET_GENRE = `SET_GENRE`,
  SET_SELECTED_MOVIE = `SET_SELECTED_MOVIE`
}

const ActionCreator = {
  setGenre: (genreName: string) => {
    return {
      type: ActionType.SET_GENRE,
      payload: genreName,
    }
  },

  setSelectedMovie: (id: number) => {
    return {
      type: ActionType.SET_SELECTED_MOVIE,
      payload: id,
    }
  }
};

const reducer = (state: State = initialState, action: Action) => {
  switch (action.type) {
    case ActionType.SET_GENRE:
      return Object.assign({}, state, {
        genre: action.payload
      });
    case ActionType.SET_SELECTED_MOVIE:
      return Object.assign({}, state, {
        selectedMovie: action.payload
      })
  }

  return state;
};

export {ActionCreator, reducer, ActionType, State};
