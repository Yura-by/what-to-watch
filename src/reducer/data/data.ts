import {Movie, Action, Comment, SendingComment} from '../../types';
import MovieAdapter from '../../adapters/movie';

enum ActionType {
  SET_MOVIES = `SET_MOVIES`,
  SET_COMMENTS = `SET_COMMENTS`,
  SET_IS_SENDING_COMMENT = `SET_IS_SENDING_COMMENT`,
  SET_IS_BAD_SENT_COMMENT = `SET_IS_BAD_SENT_COMMENT`,
  SET_IS_COMMENT_SENT = `SET_IS_COMMENT_SENT`,
}

interface State {
  allMovies: Movie[];
  comments: Comment[];
  isSendingComment: boolean;
  isBadSentComment: boolean;
  isCommentSent: boolean;
}

const initialState: State = {
  allMovies: [],
  comments: [],
  isSendingComment: false,
  isBadSentComment: false,
  isCommentSent: false,
};

const ActionCreator = {
  setMovies: (movies: Movie[]) => {
    return {
      type: ActionType.SET_MOVIES,
      payload: movies
    };
  },

  setComments: (comments: Comment[]) => {
    return {
      type: ActionType.SET_COMMENTS,
      payload: comments
    };
  },

  setIsSendingComment: (isSending: boolean) => {
    return {
      type: ActionType.SET_IS_SENDING_COMMENT,
      payload: isSending,
    }
  },

  setIsBadSentComment: (isBad: boolean) => {
    return {
      type: ActionType.SET_IS_BAD_SENT_COMMENT,
      payload: isBad,
    }
  },

  setIsCommentSent: (isSent: boolean) => {
    return {
      type: ActionType.SET_IS_COMMENT_SENT,
      payload: isSent,
    }
  }
};

const Operation = {

  loadMovies: () => (dispatch, getState, api) => {
    return api.get(`/films`)
      .then((response) => dispatch(ActionCreator.setMovies(MovieAdapter.getMovies(response.data))))
  },

  loadComments: (movieId: number) => (dispatch, getState, api) => {
    return api.get(`/comments/${movieId}`)
      .then((response) => {
        dispatch(ActionCreator.setComments(response.data));
      })
  },

  sendComment: (movieId: number, data: SendingComment) => (dispatch, getState, api) => {
    dispatch(ActionCreator.setIsSendingComment(true));
    return api.post(`/comments/${movieId}`, data)
      .then(() => {
        dispatch(ActionCreator.setIsSendingComment(false));
        dispatch(ActionCreator.setIsBadSentComment(false));
        dispatch(ActionCreator.setIsCommentSent(true));
      })
      .catch((err) => {
        dispatch(ActionCreator.setIsSendingComment(false));
        if (err.response.status === 400) {
          dispatch(ActionCreator.setIsBadSentComment(true));
        }
      });
  }
};

const reducer = (state: State = initialState, action: Action) => {

  switch (action.type) {
    case ActionType.SET_MOVIES:
      return Object.assign({}, state, {
        allMovies: action.payload
      });
    case ActionType.SET_COMMENTS:
      return Object.assign({}, state, {
        comments: action.payload
      });
    case ActionType.SET_IS_SENDING_COMMENT:
      return Object.assign({}, state, {
        isSendingComment: action.payload,
      });
    case ActionType.SET_IS_BAD_SENT_COMMENT:
      return Object.assign({}, state, {
        isBadSentComment: action.payload,
      });
    case ActionType.SET_IS_COMMENT_SENT:
      return Object.assign({}, state, {
        isCommentSent: action.payload,
      });
  }
  return state;
};

export {reducer, ActionCreator, ActionType, Operation, State, initialState};
