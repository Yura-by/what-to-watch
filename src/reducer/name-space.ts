export enum NameSpace {
  DATA = `DATA`,
  USER = `USER`,
  APP = `APP`,
}
