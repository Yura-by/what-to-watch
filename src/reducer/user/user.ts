import User from '../../adapters/user';
import {history} from '../../index';
import {AppRoute} from '../../const'

enum ActionType {
  SET_REQUIRE_AUTHORIZATION = `SET_REQUIRE_AUTHORIZATION`,
  SET_USER_DATA = `SET_USER_DATA`,
  SET_REQUEST_FLAG = `SET_REQUEST_FLAG`,
}

interface Data {
  id: number;
  email: string;
  name: string;
  avatarUrl: string;
}

interface State {
  isRequireAuthorization: boolean;
  data: Data;
  isBadRequest: boolean;
}

const initialState: State = {
  isRequireAuthorization: false,
  data: null,
  isBadRequest: false,
};

const ActionCreator = {
  setRequireAuthorization: (isRequire: boolean) => {
    return {
      type: ActionType.SET_REQUIRE_AUTHORIZATION,
      payload: isRequire
    };
  },

  setRequestFlag: (isBad: boolean) => {
    return {
      type: ActionType.SET_REQUEST_FLAG,
      payload: isBad
    }
  },

  setUserData: (userData) => {
    return {
      type: ActionType.SET_USER_DATA,
      payload: userData
    }
  }
};

interface userData {
  email: string;
  password: string;
}

const Operation = {
  checkLogin: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(ActionCreator.setUserData(new User(response.data)));
        dispatch(ActionCreator.setRequireAuthorization(false));
      })
      .catch(() => dispatch(ActionCreator.setRequireAuthorization(true)))
  },

  signIn: (userData: userData) => (dispatch, getState, api) => {
    return api.post(`/login`, userData)
      .then((response) => {
        dispatch(ActionCreator.setUserData(new User(response.data)));
        dispatch(ActionCreator.setRequireAuthorization(false));
        dispatch(ActionCreator.setRequestFlag(false));
        history.push(AppRoute.ROOT);
      })
      .catch((err) => {
        if (err.response.status === 400) {
          dispatch(ActionCreator.setRequestFlag(true));
        }
      })
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_REQUIRE_AUTHORIZATION:
      return Object.assign({}, state, {
        isRequireAuthorization: action.payload
      });
    case ActionType.SET_REQUEST_FLAG:
      return Object.assign({}, state, {
        isBadRequest: action.payload
      });
    case ActionType.SET_USER_DATA:
      return Object.assign({}, state, {
        data: action.payload
      });
  }

  return state;
};

export {ActionType, ActionCreator, reducer, Operation, initialState};
