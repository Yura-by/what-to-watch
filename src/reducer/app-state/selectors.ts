import {createSelector} from 'reselect';

import {Store} from '../../types';
import {NameSpace} from './../name-space';

import {getAllMovies} from '../data/selectors'

import {getUniqItems} from '../../utils';
import {Genre} from '../../const';
import {Movie} from '../../types';

const getActiveGenre = (state: Store) => {
  return state[NameSpace.APP].genre;
};

const getAllGenres = createSelector(
  getAllMovies,
  (movies) => {
    const allGenres: string[] = movies.map((it) => it.genre);
    return getUniqItems<string>(allGenres);
  }
);

const getMoviesByType = createSelector(
  getActiveGenre,
  getAllMovies,
  (genre: string, allMovies: Movie[]) => {
    if (genre === Genre.ALL_GENRES) {
      return allMovies;
    }
    return allMovies.filter((it) => it.genre === genre);
  }
);

const getSelectedMovieId = (state: Store) => {
  return state[NameSpace.APP].selectedMovie;
};

const getSelectedMovie = createSelector(
  getSelectedMovieId,
  getAllMovies,
  (id, movies) => {
    return movies.find((it) => it.id === id);
  }
)

export {getAllGenres, getAllMovies, getActiveGenre, getMoviesByType, getSelectedMovieId, getSelectedMovie};

