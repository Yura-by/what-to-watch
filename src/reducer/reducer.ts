import {combineReducers} from 'redux';

import {NameSpace} from '../reducer/name-space';
import {reducer as data} from './data/data';
import {reducer as user} from './user/user';
import {reducer as app} from './app-state/app-state';

export default combineReducers({
  [NameSpace.DATA]: data,
  [NameSpace.USER]: user,
  [NameSpace.APP]: app,
});
