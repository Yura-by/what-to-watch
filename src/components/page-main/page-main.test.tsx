import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Movies} from '../../mock/movies';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import PageMain from './page-main';

const mockStore = configureStore([]);

import {NameSpace} from '../../reducer/name-space';

import {initialState as dataInitialState} from '../../reducer/data/data';
import {initialState as userInitialState} from '../../reducer/user/user';
import {initialState as appInitialState} from '../../reducer/app-state/app-state';

const store = mockStore({
  [NameSpace.APP]: appInitialState,
  [NameSpace.DATA]: {...dataInitialState, allMovies: Movies},
  [NameSpace.USER]: userInitialState,
});

it(`PageMain snapshot test`, () => {
  const tree = renderer
    .create(<Provider store={store}>
      <PageMain
        allMovies={Movies}
        onSelectMovie={jest.fn()}
      />
    </Provider>).toJSON();

  expect(tree).toMatchSnapshot();
});
