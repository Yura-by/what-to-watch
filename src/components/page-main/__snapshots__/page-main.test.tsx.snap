// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageMain snapshot test 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Pulp Fiction"
        src="https://htmlacademy-react-3.appspot.com/wtw/static/film/background/Pulp_Fiction.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="sign-in.html"
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="Pulp Fiction"
            height="327"
            src="https://htmlacademy-react-3.appspot.com/wtw/static/film/poster/Pulp_Fiction.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Pulp Fiction
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Crime
            </span>
            <span
              className="movie-card__year"
            >
              1994
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item catalog__genres-item--active"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Crime
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Action
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Thriller
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Adventure
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Comedy
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Drama
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Fantasy
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Pulp Fiction"
              height="175"
              src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/pulp-fiction.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
            >
              Pulp Fiction
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Midnight Special"
              height="175"
              src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/midnight-special.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
            >
              Midnight Special
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Shutter Island"
              height="175"
              src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/shutter-island.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
            >
              Shutter Island
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Once Upon a Time in, America"
              height="175"
              src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/Once_Upon_a_Time_in_America.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
            >
              Once Upon a Time in, America
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Moonrise Kingdom"
              height="175"
              src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/moonrise-kingdom.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
            >
              Moonrise Kingdom
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="What We Do in the Shadows"
              height="175"
              src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/what-we-do-in-the-shadows.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
            >
              What We Do in the Shadows
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Aviator"
              height="175"
              src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/aviator.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
            >
              Aviator
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Fantastic Beasts: The Crimes of Grindelwald"
              height="175"
              src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/fantastic-beasts-the-crimes-of-grindelwald.jpg"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
            >
              Fantastic Beasts: The Crimes of Grindelwald
            </a>
          </h3>
        </article>
      </div>
      <div
        className="catalog__more"
      >
        <button
          className="catalog__button"
          onClick={[Function]}
          type="button"
        >
          Show more
        </button>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          © 2020 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
